File: frontend/src/services/cartService.js
import api from './api';

/**
 * Service Layer – Cart API calls.
 * All endpoints require a valid CUSTOMER JWT (handled by api.js request interceptor).
 * Every write operation returns the full updated CartResponse so the caller
 * does not need a separate GET to stay in sync.
 */

/**
 * Fetch the current cart state for the authenticated customer.
 * Returns an empty cart (items: [], totalAmount: 0, itemCount: 0) when no cart exists.
 * @returns {Promise<CartResponse>}
 */
export const getCart = () =>
  api.get('/api/cart').then((res) => res.data);

/**
 * Add a product to the cart, or increment its quantity if already present.
 * @param {number} productId
 * @param {number} [quantity=1]
 * @returns {Promise<CartResponse>}
 */
export const addToCart = (productId, quantity = 1) =>
  api.post('/api/cart/items', { productId, quantity }).then((res) => res.data);

/**
 * Update the quantity of a specific cart item.
 * @param {number} cartItemId  - the cart item's primary key
 * @param {number} quantity    - new quantity (≥ 1)
 * @returns {Promise<CartResponse>}
 */
export const updateCartItem = (cartItemId, quantity) =>
  api.put(`/api/cart/items/${cartItemId}`, { quantity }).then((res) => res.data);

/**
 * Remove a single item from the cart.
 * @param {number} cartItemId
 * @returns {Promise<CartResponse>}
 */
export const removeCartItem = (cartItemId) =>
  api.delete(`/api/cart/items/${cartItemId}`).then((res) => res.data);

/**
 * Remove all items from the cart (called after a successful order placement).
 * @returns {Promise<void>}
 */
export const clearCart = () =>
  api.delete('/api/cart');

File: backend/src/main/java/com/urbanfresh/service/CartService.java
package com.urbanfresh.service;

import com.urbanfresh.dto.request.AddToCartRequest;
import com.urbanfresh.dto.request.UpdateCartItemRequest;
import com.urbanfresh.dto.response.CartResponse;

/**
 * Service Layer – Contract for cart management operations.
 * All methods are scoped to the authenticated customer identified by their email.
 */
public interface CartService {

    /**
     * Retrieve the customer's current cart.
     * Returns an empty cart response when the customer has no cart yet.
     *
     * @param customerEmail email extracted from the JWT
     * @return current CartResponse with items and totals
     */
    CartResponse getCart(String customerEmail);

    /**
     * Add a product to the cart, or increment its quantity if it is already present.
     * Rejects products that are out of stock.
     *
     * @param customerEmail email extracted from the JWT
     * @param request       product ID and quantity to add
     * @return updated CartResponse after the addition
     */
    CartResponse addToCart(String customerEmail, AddToCartRequest request);

    /**
     * Update the quantity of an existing cart item.
     * Only the owning customer may update their own items.
     *
     * @param customerEmail email extracted from the JWT
     * @param cartItemId    ID of the cart item to update
     * @param request       new quantity
     * @return updated CartResponse after the change
     */
    CartResponse updateCartItem(String customerEmail, Long cartItemId, UpdateCartItemRequest request);

    /**
     * Remove a single item from the cart.
     * Only the owning customer may remove their own items.
     *
     * @param customerEmail email extracted from the JWT
     * @param cartItemId    ID of the cart item to remove
     * @return updated CartResponse after the removal
     */
    CartResponse removeCartItem(String customerEmail, Long cartItemId);

    /**
     * Remove all items from the customer's cart (e.g. after successful order placement).
     *
     * @param customerEmail email extracted from the JWT
     */
    void clearCart(String customerEmail);
}
